### ISSUE TITLE
I need the opposite of exponential slider slope & ADSR discussion

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#80]
Could you implement Parameter::scalings::LOGARITHMIC or whatever is the opposite of exp?

**Comment by RobinSchmidt** [comment]
depends on how you define opposite. if it means "inverse function", the yes, logarithm. but it could also mean just taking the negative or 1-exp(a*x) - which would look like a RC loading curve. for what kind of parameter is this?

**Comment by elanhickler** [comment]
I needed it because my parameter needs reversing: 1-v, so i need the high end of the slider to to have very small incrementing values rather than the low end of the slider.

But I think I may not need this anymone, I am so frustrated with my ADSR, I want to give up. Gonna look at your analog (overly compelx) adsr.

**Comment by elanhickler** [comment]
I'm going to do the one pole trigger signal approach.

Two pole approach!

filerCutoff = onepoleA.getSample(1/newTimingValue);
onePoleB.setCutoff(filterCutoff);
output_value = onepoleB.getSample(valueToMoveTo);

**Comment by RobinSchmidt** [comment]
> I am so frustrated with my ADSR, I want to give up

why? what's wrong with it? that you can't get the timing decoupled from the shape?

> Gonna look at your analog (overly compelx) adsr.

i think, you can just ignore the features you don't need. the default settings should behave neutral, i.e. if you don't call any of the setPeakLevelByVel, etc. stuff, the deafult settings make it behave as if these parameters were not there in the first place.

**Comment by RobinSchmidt** [comment]
> my parameter needs reversing: 1-v

that would be 1-exp(-decay*t) then, i.e. one minus an exponential decay function - which gives and exponential saturation curve - like an RC loading curve

**Comment by RobinSchmidt** [comment]
```
filerCutoff = onepoleA.getSample(1/newTimingValue);
onePoleB.setCutoff(filterCutoff);
output_value = onepoleB.getSample(valueToMoveTo);
```
that looks more like smoothing over time to me rather than an instantaneous mapping function

**Comment by elanhickler** [comment]
Right now I am using rosic one pole and setting cutoff for each state to get the right timing. I do "if target value was reached, go to next stage" so of course timing is all off.

Then i did "if time passed, go to next stage" except for release because I don't want to cut off release tails, i want to let the filter decay.

Now the problem is... lets say your attack was slow and only for a split second, so your actual value is something low like .1. It takes FOREVER for the filter get to near 0, because actual value and target value different is so little. Your rosic::OnePole must be really bad for timing.

What should I do?

> I am so frustrated with my ADSR, I want to give up

> > why? what's wrong with it? that you can't get the timing decoupled from the shape?

exactly. If I try to do anything more complex than a basic linear ADSR, all my efforts go to s***

**Comment by elanhickler** [comment]
Hmm, maybe what I really need is a curve function, so I never worry about timing.

**Comment by elanhickler** [comment]
Ok, I just improved my linear ADSR so it does everything perfectly. Timed-based looped, value-reaching-based loop (I like this untimed behavior for when doing modulation / sound design, but perfectly timed behavior is good for music).

**Comment by elanhickler** [comment]
I guess maybe I just want a modulatable version of your breakpoint ADSR, using "analog" shape.... does the shape only update on attack/reset? How modulatable is it?

**Comment by elanhickler** [comment]
ok so it's not modulatable at all!

> ```
> SmoothedValue:
> -see class juce::LinearSmoothedValue
> -generalize to an update equation: x[n+1] = a*x[n] + b
> -allows for linear and exponential smoothing and anything between
> -general term is: 
>  x[n] = a^n * x[0] + b * sum_{k=0}^{n-1} a^n 
>       = a^n * x0   + b * n * a^n 
>       = a^n * (x0 + b*n)
> -maybe we can have a kind of "exponentiality" parameter p between 0..1
>  with 0, we have linear scaling (a = 1), with 1 exponential (b = 0)
>  maybe a = 1-p, b = p * c (for some suitable c, maybe 1/N (?) where N 
>  is the desired number of updates to reach the target)
> ```

I guess this is what i want!

**Comment by elanhickler** [comment]
x[n] = a^n * x[0] + b * sum_{k=0}^{n-1} a^n 
      = a^n * x0   + b * n * a^n 
      = a^n * (x0 + b*n)

What are the three equations?

and what is going on here: `sum_{k=0}`, doesn't look like C++

I could try to implement an ADSR based on this.

**Comment by RobinSchmidt** [comment]
yeah - it's LaTeX notation. i don't know how to write down equations in plain text otherwise

**Comment by RobinSchmidt** [comment]
...i'll come back to this

**Comment by RobinSchmidt** [comment]
https://en.wikipedia.org/wiki/LaTeX

**Comment by RobinSchmidt** [comment]
it's a typesetting system that includes all the good stuff for mathematics:

https://en.wikibooks.org/wiki/LaTeX/Mathematics

in RSLib, i've used that for the documentation. if you generate a doxygen documentaion there, you'll get a nice pdf document with proper math typesetting. unfortunately, it makes the code comments less readable - you have all that additional distracting markup there. ...sooo you can get a nice pdf documentation but the documentation in t h/cpp files is uglified. i'm not sure what to do in rapt

**Comment by RobinSchmidt** [comment]
btw. - a very strange observation that i made: when i read texts that include math, the appearance/typesetting seems to make a difference for, how hard or easy it is to understand. if the typesetting is aesthetically pleasing, the content gets easier to understand

**Comment by elanhickler** [comment]
check out these curves: http://www.mathopenref.com/graphfunctions.html?fx=x^(a/x)&gx=x^c&hx=x^(x/b)^(x*b)&xh=1&xl=0&yh=1&yl=0&ah=2&a=0.22340425531914893&bh=10&bl=-10&b=1.8&c=0.3

**Comment by RobinSchmidt** [comment]
> check out these curves

do you think they are useful? i could make my Parameter class more flexible by allowing the user to pass in a kind of ValueMapper object instead of providing a fixed collection of enumerated mapping functions

**Comment by RobinSchmidt** [comment]
like: `setValueMapper(ValueMapper *newMapper)`.

then, you could subclass ValueMapper and do any mapping function you want. that would be the strategy pattern:

https://sourcemaking.com/design_patterns/strategy

...which is, btw., the same idea i want to use for making the gui customizable. there, you could then do something like `mySlider->setPainter(mySliderPainter)` ....or something

**Comment by elanhickler** [comment]
those curves are useful for envelope generators, no idea what I will need for mapping, most likely just log/exp curves.

**Comment by elanhickler** [comment]
i just went back to my old ADSR code and actually made something really awesome with it.

![image](https://user-images.githubusercontent.com/9116575/30997298-b6535038-a47b-11e7-8636-df8965789b9a.png)

The controls need to be better placed and sized, but it's pretty much finished. It would be nice for you to create the kind of linear smoother/shapable smoother you were talking about above, would love to have that for an ADSR, but for now I will close this issue.

**Comment by elanhickler** [comment]
https://hearthis.at/soundemote/basicadsr-feedback-looping/

This is the direct output of the envelope generator, used in such a way that it behaves like an oscillator. This is a test of a feedback-shaper-style ADSR in loop mode where the feedback chaotically interacts with the looping mechanism. When using lower ADSR times, the feedback helps to create very natural attack and decay shapes, from smooth/organic to extremely snappy. At first you are hearing an LFO controling envelope timing to create the modulation. At the end I turn off the LFO to modulate timings manually.

**Comment by elanhickler** [comment]
f*** my ADSR is still locking on to divisions of the samplerate.

My ADSR is basically trying to be a 1 dimensional version of your RayBouncer except that sustain represents a movable boundary, and the sustain value should be held once reached unless it's i looping mode then it should continue bouncing.

I need help with this. I need to commission you to get my ADSR working more like your RayBouncer.

It's really simple:

1. Start by creating a linear ADSR where value is incremented and decremented based on time it takes to get from currentValue to 1 or 1 to sustainAmp value.

Example: 
`attackIncrement = (1-currentValue) * secondsPerSample / attackTime`
decayIncrement = (currentValue-sustainAmp) * secondsPerSample / decayTime;

Basically a linear smoother that has a starting value and target value and wants to reach the target value in the time given.

2. We need Attack/Decay/Sustain/Release stages. Attack stages goes from currentValue to 1. Once 1 is earched, we go from 1 to sustainAmp value. If looping, we go back to 1 based on Attack time. If not looping, hold sustainAmp value until "isSustaining" is set to false, or whatever... release callback. From release, that's easy, just go from currentValue to 0.

3. If we are in loop mode and we have to calculate multiple reflections per sample, limit to calculating 1 reflection, and after that, set value to random number between 0 and 1.

My code is so close to working, man it would be nice if you just met me online, you can charge me hourly, and we go over my code and see if we can find an easy fix, see what I'm doing wrong.

**Comment by RobinSchmidt** [comment]
ok, maybe tomorrow in the late afternoon early evening? i'm a bit tired at the moment. but what exactly is the problem? is there a problem only when you are looping and have multiple reflections per sample? how does such a situation even occur? you must be running the env at insane speeds and or have near-zero transition times.

btw. the bill is currently at 543 with 197 for the cycle-mark stuff, so i think, it's not yet time for a new invoice

**Comment by elanhickler** [comment]
ok, yes tomorrow is fine.

the problem is that the frequency is locking on to divisions of the samplerate, yes I'm running it at insane speeds, it's essentially supposed to be an oscillator with a sustain level. 1 dimensional Ray Bouncer with an adjustable wall called Sustain 😄

**Comment by RobinSchmidt** [comment]
ok, i'm making a 1D raybouncer with two adjustable wall called floor and ceil...

**Comment by RobinSchmidt** [comment]
ok..there's a class rsRayBouncer1D in rapt. it happily bounces back and forth between floor and ceiling:
![image](https://user-images.githubusercontent.com/1833099/31561285-aa0a7940-b057-11e7-9ec1-df0636605403.png)

for the time being, i just do multiple reflection, if necessarry

**Comment by RobinSchmidt** [comment]
with an increment of 3.1, we get multiple reflections per sample which results in weird patterns:
![image](https://user-images.githubusercontent.com/1833099/31561444-30b42d1a-b058-11e7-9718-08d08d4b03a3.png)

**Comment by elanhickler** [comment]
It looks like you're getting values below 0. If that only happens at high frequency, that is a good time to call `makeRandomDecision(double * vaue)` (a function that takes a value and changes it randomly between 0 and 1).

**Comment by RobinSchmidt** [comment]
as said, both, floor and ceiling are adjustable. i had the floor set to 0.2 in the plot (and the ceiling at 0.7)

**Comment by RobinSchmidt** [comment]
i think, if you want random decisions, i should make that optional (like, having a reflection mode parameter)

**Comment by elanhickler** [comment]
yeah, I don't exactly know what would work, but maybe... limiting reflections to 2 per sample and then random decision after that might work..?

**Comment by RobinSchmidt** [comment]
maybe not a random decision but some deterministic formula based on floor, ceil and value (and maybe increment)... that might come out pseudo-random anyway

**Comment by elanhickler** [comment]
right, yes, that would work too. Here's my code for shapes.

```cpp
double FeedbackADSR::calculateNextVaue(double inc, Shape shape, double fbAmount)
{
	inc = jmax(1.e-6, inc);

	double out;
	switch (shape)
	{
	case Shape::EXP:
		out = inc + currentValue*fbAmount + fbAmount*r*.01 + r*.01;
		break;
	case Shape::LOG:
		out =  inc + (1-currentValue)*fbAmount + fbAmount*r + r*.001;
		break;
	case Shape::SIGMOID:
		out =  inc + (1-currentValue)*currentValue*fbAmount + fbAmount*r*.0001;
		break;
	case Shape::POW:
		out =  inc + currentValue*currentValue*fbAmount + fbAmount*r*.01 + r*.0001;
		break;
	}

	// level += A*level^2 + B*level + C

	//out = jmin(out, 0.5);

	return out;
}
```

**Comment by RobinSchmidt** [comment]
what is `r`?

**Comment by elanhickler** [comment]
sorry, im writing you a simplified version, one sec. r is random value / noise. I added noise to the entire system with that in my attempt to make things more interesting. :)

**Comment by elanhickler** [comment]
```cpp
void incrementValue(double * currentValue, double increment, Shape shape, double shapeAmt)
{
	switch (shape)
	{
	case Shape::LIN:
		*currentValue += inc;
	case Shape::EXP:
		*currentValue += inc + currentValue*shapeAmt;
	case Shape::LOG:
		*currentValue += inc + (1-currentValue)*shapeAmt;
	case Shape::SIGMOID:
		*currentValue += inc + (1-currentValue)*currentValue*shapeAmt;
	case Shape::POW:
		*currentValue += inc + currentValue*currentValue*shapeAmt;
	}
}
```

**Comment by elanhickler** [comment]
a simplified formula is this, basically allowing any amounts of all shapes:

`level += A*level^2 + B*level + C`

and you can go crazier and do

`level += A*level^pow + B*level + C`

having adjustable exponentialness.

but for my purposes/control scheme I'm saving CPU by only doing the needed math for the desired shape.

**Comment by RobinSchmidt** [comment]
the first one looks actually simple enough. i the switch really faster?

**Comment by elanhickler** [comment]
switch statements are similar speeds or faster than if else: https://stackoverflow.com/questions/1028437/why-switch-case-and-not-if-else-if

**Comment by RobinSchmidt** [comment]
uuuh...your formula seems to ultra sensitive to the feedback. a little too much and it gets stuck at the ceiling...

**Comment by elanhickler** [comment]
what? you must be getting reverse feedback. Are you incrementing/decrementing correctly and not JUST incrementing?

**Comment by RobinSchmidt** [comment]
```cpp
  void increment()
  {
    x += dx + shape*x;
  }
```

**Comment by elanhickler** [comment]
and you increment TOWARD 0 in the case of bouncing off the ceiling?

**Comment by RobinSchmidt** [comment]
my reflection code is a bit complex due to supporting multiple reflections, the whole state update procedure currently goes like this:
```cpp
  void increment()
  {
    x += dx + shape*x;
  }

  void reflect()
  {
    while(x > ceil || x < floor)
    {
      if(x > ceil)
      {
        T over = x - ceil;
        x = ceil - over;
        dx = -inc;
      }
      if(x < floor)
      {
        T under = floor - x;
        x = floor + under;
        dx = inc;
      }
    }
  }

  inline T getSample()
  {
    T out = x;
    increment();
    reflect();
    return out;
  }
```

**Comment by elanhickler** [comment]
wait a second, we need separate increment/decrement times. you only have function "set increment"?

**Comment by RobinSchmidt** [comment]
ahh...ok - yes. i currently use the same value either positive or negative. but that's a quick change...and should not really have much to do with feedback sensitivity/instability

**Comment by elanhickler** [comment]
ok, going to pull your changes and see if I can figure out the feedback problem.

**Comment by RobinSchmidt** [comment]
i just checked in an update with separate increment and decrement

**Comment by elanhickler** [comment]
hmm, well I'm noticing that increment speed affects shape, which shouldn't be the case... my envelope didn't do that. Will keep testing.

**Comment by elanhickler** [comment]
OH...ohohohoh

**Comment by RobinSchmidt** [comment]
what kind of testbed do you actually use? dis you quickly turned that into a plugin?

**Comment by elanhickler** [comment]
yeah, im testing in my jerobeammushroom. Just changed my ADSR to use your rayboucer1d.

**Comment by elanhickler** [comment]
Your 1D RayBouncer is exhibiting frequency locking similar to my envelope. Also I just tested y-he TyrellN6 and it also has the same issue........................... WHAT

**Comment by RobinSchmidt** [comment]
what exactly do you mean by frequncy locking?

**Comment by elanhickler** [comment]
as you change a parameter (thereby effectively changing speed of reflections), the speed of reflections is more and more quantized as you get up in frequency, as you approach nyquist.

**Comment by RobinSchmidt** [comment]
> Agh, another problem with 1D Raybouncer, as you change attack toward near instant, you get this issue (decay doesn't start at ceiling)

ah - i think, i know why this is. the code (implicitly) assumes, that after hitting the wall, it gets reflected with attack-speed rather than decay speed. to solve this, more care has to be taken in the reflection code

> You only get frequency locking issues when the shape is not perfectly triangle.

ok, that's an important hint. i think, it means that we are not taking into account curving in the reflection code, but should....somehow...but how?

**Comment by elanhickler** [comment]
> ah - i think, i know why this is. the code (implicitly) assumes, that after hitting the wall, it gets reflected with attack-speed rather than decay speed. to solve this, more care has to be taken in the reflection code

hmm, see if this solves the problem, or make the change and I'll test again when I can.

> ok, that's an important hint. i think, it means that we are not taking into account curving in the reflection code, but should....somehow...but how?

I'm thinking we can't use incremental math, and we have to use explicit math, i.e. modulatable version of your breakpoint envelope, except that I would want to keep the behavior where changing curve also changes reflection speed (i guess, as an option)

**Comment by RobinSchmidt** [comment]
>> after hitting the wall, it gets reflected with attack-speed rather than decay speed. 
> hmm, see if this solves the problem, or make the change and I'll test again when I can.

ok, done (i think). this is the relevant change (and a similar one for the floor):

```cpp
      if(x > ceil)
      {
        T over = x - ceil;
        over *= dec/inc;      // assume decay-velocity from the moment of hitting the ceiling
        x = ceil - over;
        dx = -inc;
      }
```

**Comment by RobinSchmidt** [comment]
> I'm thinking we can't use incremental math, and we have to use explicit math

actually, at each step, we do:

x += v(x)

where the velocity can be considered as a function of x. in the case of a linear shape, we just have 

v(x) = a

for some constant a, if we add some of x to that, we get

v(x) = a + b*x

the position at any instant is then given as a sum:

x(n) = x(0) +  sum_n v_n

if we can derive a closed form for this sum, we may be able to account for velocity changes (i.e. accelerations) within the reflection. ...i'm a bit shaky about all of this - just some brainstorming

**Comment by RobinSchmidt** [comment]
this sum to compute position is actually similar to an integral over the velocity and can perhaps be approximated by this integral (integrals of polynomials are easy).

**Comment by RobinSchmidt** [comment]
hmm, the sums should actually be easy to solve too - we need ingredients of the form:

http://www.wolframalpha.com/input/?i=sum+over+n+of+n

http://www.wolframalpha.com/input/?i=sum+over+n+of+n%5E2

(the partial sums)

**Comment by RobinSchmidt** [comment]
but as said...take all of that with multiple grains of salt

**Comment by elanhickler** [comment]
I'd be willing to pay for addition work in this area if my product release is successful. For now I just need to release a product. I don't need you to do anything desperately, but getting the smoother done would be nice. Breakpoint bug fixes too.

**Comment by elanhickler** [comment]
Hey Robin, you know we could possibly simulate attack/decay times by taking a tri-saw wave (where it can be shaped to be saw to tri), and just... correlating certain parameters with certain other parameters, I mean...

imagine a triangle, skew it a little bit so it begins to approach a sawtooth shape, boom, now you have a short attack and a longer decay. Want to increase decay time? You have to lower frequency, but keep the triangle's attack shape/time the same, so decrease frequency AND increase the "pulse width" toward one side, you follow?

To create curves, now you just do feedback frequency modulation.

Now, obviously it would require knowing how to correlate parameters in various ways, and I wouldn't know how to deal with turning this into a full ADSR, hmm... maybe this idea would only work if my goal was to create an AD envelope, but it would make one hell of a cool oscillator/AD env.

**Comment by elanhickler** [comment]
hmm, but this idea would not even be able to have separate attack/decay shapes, because if you tried to switch curves for attack or decay, you could be in the decay portion within a few sample, it has all the same potential issues as anything else, unless you just want a really weird A/D envelope.

**Comment by RobinSchmidt** [comment]
>  obviously it would require knowing how to correlate parameters in various ways

lets call a: attack-time, d: decay-time, p: period (overall length of a cycles). i think then this just means that we should keep the invariant a + d = p at all times. however, i'm a bit skeptical about using frequency modulation to control the shape. wouldn't that potentially mess up the whole waveform shape? hmm don't know.

however, i have though some more about finding explicit formulas for the envelope value in terms of sample index n. at least for the case with only linear feedback, it's not too hard actually. i added a comment to the code file:
```
The update equation in increment() is:

x += dx + shape*x;

This is a recurrence relation of the form:

x[0] = s                 // start value
x[n] = b*x[n-1] + a      // b = 1 + shape, a = dx

the first few term of which come out as:

x[1] = b*x[0] + a                       = b * s +     a      
x[2] = b*x[1] + a = b*(b*s + a) + a     = b^2*s + b * a
x[3] = b*x[2] + a = b*(b^2*s + b*a) + a = b^3*s + b^2*a

from which we see the genral pattern:

x[n] = b^n * s + b^(n-1) * a

...hmm, but wolfram alpha says:
x(n) = (a b^n - a + s b^(n+1) - s b^n)/(b-1)
for this input:
RSolve[{x[n] == b x[n-1] + a,x[0]==s}, x[n], n]
```

it's a bit strange that my result is different from wolfram's:

http://www.wolframalpha.com/input/?i=RSolve%5B%7Bx%5Bn%5D+%3D%3D+b+x%5Bn-1%5D+%2B+a,x%5B0%5D%3D%3Ds%7D,+x%5Bn%5D,+n%5D

...i guess we could just try both formulas in an experimental code testbed

**Comment by RobinSchmidt** [comment]
ahhh - i just see my error! i forgot to add the final "a" in the last column. soo...ok, yeah, my pattern is wrong, wolfram is right.

**Comment by RobinSchmidt** [comment]
with this formula in hand, we could potentially solve for the sub-sample time-instant, when the value hits the wall (just plug in the wall-value for the lhs and solve for n). ...and with this, it might be possible to compute subsample accurate reflections, taking into account the feedback

**Comment by RobinSchmidt** [comment]
if this all works out, a similar approach could be used to include quadratic feedback as well. ...the formulas will undoubtly become messier, though

**Comment by RobinSchmidt** [comment]
>  I switched to using polyBLEP waveforms.

for what? where did you get the code from for them?

**Comment by RobinSchmidt** [comment]
> with this formula in hand, we could potentially solve for the sub-sample time-instant, when the value hits the wall (just plug in the wall-value for the lhs and solve for n - s,a,b are just known numbers)

and the result is:

http://www.wolframalpha.com/input/?i=Solve%5Bw+%3D%3D+(a+b%5En+-+a+%2B+s+b%5E(n%2B1)+-+s+b%5En)%2F(b-1),+n%5D

**Comment by elanhickler** [comment]
> wouldn't that potentially mess up the whole waveform shape? hmm don't know.
that's a desired effect.

**Comment by elanhickler** [comment]
BTW, if you get this triangle bouncing oscillator working perfectly, it's going to be one hell of an awesome oscillator. I think there's so many uses for this kind of thing. Morphable envelopes, oscillators, chaos generators...

**Comment by RobinSchmidt** [comment]
yes, but i mean, mess it up in a way that doesn't really give you distinguished attack- and release stages anymore?. hmm...well, i think, that would be the case for regular two-osc fm (it could wiggle up/down multiple times per cycle). not sure about what happens in feedback fm. although, with a sine, it tends to shape into a saw, maybe something similar happens with a trisaw

**Comment by elanhickler** [comment]
you're not sure what happens in feedback FM? just use my mushroom vst (you won't be able to get it to build at the moment), but if you had that, you could experiment with any kind of feedback you want.

**Comment by RobinSchmidt** [comment]
> if you get this triangle bouncing oscillator working perfectly, it's going to be one hell of an awesome oscillator.

ok, so what is still missing right now? i assume, when you say triangle, you mean, "without feedback", i.e. without curving?

**Comment by elanhickler** [comment]
oh, i mean with curving. a simple triangle oscillator is not so useful.

**Comment by elanhickler** [comment]
btw, if you added repelling/force to the walls (so the oscillator slows down as it approach as wall) that would be a form of antialiasing.

**Comment by RobinSchmidt** [comment]
ok, i will try to implement sub-sample accurate reflections with curving based on the formulas, i posted above

**Comment by elanhickler** [comment]
ok, you can add that to my bill provided it's not a long arduous process.

**Comment by RobinSchmidt** [comment]
it took me 1:45h. i hope it works. i did some experiments which look not that bad. i still have the instability/sensitivity problem with the feedback/shape parameter. i can use only small values...maybe it should be made dependent on the sample rate

**Comment by RobinSchmidt** [comment]
hmm...somehow it still looks wrong..

**Comment by elanhickler** [comment]
ok i gotta try this. You know this 1D or even 2D and your particle bouncer in 3D, all these things could be used to create perfectly antialiased oscillators I believe. I think in this case you sacrifice perfect tuning for perfect antialiasing. I think you will have to implement a repulsion effect for the walls to further this idea of perfectly shaped oscillators, but I think what you created now can demonstrate the concept. I'll see if I can demonstrate it now with your 1D bouncer.

**Comment by elanhickler** [comment]
I'm getting build errors, unresolved ReflectCurved / ReflectLinear

**Comment by elanhickler** [comment]
Ok i fixed the build errors in my own way, not sure how you want to do it. Also I am limiting the while statement for reflections otherwise it easily causes infinite hang.

This is very promising! I need the S curve shape to demonstrate antialiasing though! Can you enable that shape easily? Or does that require a whole new reflection formula?

**Comment by RobinSchmidt** [comment]
> does that require a whole new reflection formula?

that's with x^2 term, right? yes, this will need another formula. i just tried to find it via wolfram alpha, but it gives up. that will probably be impractical to do in the same way

**Comment by elanhickler** [comment]
Would it be easier to implement a wall force? You've been doing some research into that for 3D ray bouncer.

**Comment by RobinSchmidt** [comment]
hmm...i think, something like that would indeed be easier to implement - but a totally different concept. for example, look at this graph:
https://www.desmos.com/calculator/guj25rseef
if we use this as force (pointing toward the origin), it could oscillate between -1 and 1. with high exponents, it would be essentially force-free near the origin (so it would move in a straight line). the lower the exponent, the more rounded the shape. if we combine this with another force that is either downward or upward only, we may get asymmetrical waves.

**Comment by RobinSchmidt** [comment]
we could also move the force function away from the origin (also dynamically), such that the signal is not necessarily centered at the zero line

**Comment by RobinSchmidt** [comment]
or maybe more like that:
https://www.desmos.com/calculator/wdyax8pxuo
with p=1, we'd get sinusoidal motion (the particle would be connected to 0 by a hooke-spring-like force), with higher p, we would approach a triangle wave

**Comment by elanhickler** [comment]
I think you will get shapes similar to this: https://www.desmos.com/calculator/7h9mknbv3q (i have a recursive formula version of this as well, the xoxos oscillator)

I was hoping for more of a triangle with variably rounded tops that I guess would also transition to sine given enough wall force, I guess it would end up being very similar to PolyBLAMP (polynomial band limited ramp). rather than what you are showing me which looks just as good as amplitude-crossfading mixing triangle and sine, here's my crappy mspaint attempt:

![image](https://user-images.githubusercontent.com/9116575/31694764-f8f35940-b35a-11e7-9412-58f9cddbfc47.png)

But this may be getting a bit off topic. I just need the S curve style shape somehow. But... I don't need it now, I'm going to use my own ADSR for now.

**Comment by elanhickler** [comment]
OMG! YES! If you mix in linearness with the xoxos oscillator, you get exactly this!!! This could potentially beat PolyBLEP and PolyBLAMP, 100% free of aliasing, perfectly oscillator.

Tri / Sine: https://www.desmos.com/calculator/letu72yqqt
Saw / Sine: https://www.desmos.com/calculator/7i9xuhvpor

**Comment by RobinSchmidt** [comment]
> I think you will get shapes similar to this: https://www.desmos.com/calculator/7h9mknbv3q (i have a recursive formula version of this as well, the xoxos oscillator)

this is cool! can i use that in my library? ...i think you already said that. this is the formula based on these ellipse graphs that you posted somewhere else, right?

**Comment by RobinSchmidt** [comment]
i think, cos(t), sin(t) can easily be created recursively by a rotating complex phasor. i actually already have code for that in RSLib (class rsComplexExponentialIterator). 

with that, we could even include damping! ...but then we would somehow have to drive it with an external signal. ...i think, with this idea, a filter resonance waveform based on these equations could be built :-O

**Comment by elanhickler** [comment]
I guess you can use it, xoxos didn't say anything about needing permission. 

See Soundemote\Shared\se_framework\ElanSynthLib\SynthesizerComponents\oscillator\XoxosOscillator.h for the recusive version of the oscillator inside an oscillator class.

**Comment by RobinSchmidt** [comment]
where was this thread where you posted the pics with the ellipses?

**Comment by RobinSchmidt** [comment]
...looks like the C parameter could use bipolar exp-scaling. i've implemented it for the parameter but it doesn't work yet because i think the slider class needs to be updated, too. work for tomorrow

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/31754989-406b668c-b450-11e7-89ed-15a84b348ffb.png)

**Comment by RobinSchmidt** [comment]
yes! this! but do you still have the links to our discussion and xoxos' kvr post about this? i'd like to learn more about how this was derived. i have included an experiment with these formulas into the codebase (in the (test-section), just so i have the formulas in a safe place

**Comment by RobinSchmidt** [comment]
how about using feedback modulation for the a,b,c parameters? or even using a pair of these with self- and cross-feedback modulation? should give some nice chaos for the timbre, i guess. but at the same time stable pitch, because the basic rotation frequencies are unaffected

**Comment by elanhickler** [comment]
ooooo yes I could try that!
